agent/consul/acl_endpoint.go

```diff
@@ -1412,3 +1476,559 @@ func (a *ACL) RoleResolve(args *structs.ACLRoleBatchGetRequest, reply *structs.A
+		return respErr
+	}
+
+	if _, rule, err := a.srv.fsm.State().ACLBindingRuleGetByID(nil, rule.ID); err == nil && rule != nil {
+		*reply = *rule
+	}
+
+	return nil
+}
+
+func (a *ACL) BindingRuleDelete(args *structs.ACLBindingRuleDeleteRequest, reply *bool) error {
+	if err := a.aclPreCheck(); err != nil {
+		return err
+	}
+
+	if !a.srv.InACLDatacenter() {
+		args.Datacenter = a.srv.config.ACLDatacenter
+	}
+
+	if done, err := a.srv.forward("ACL.BindingRuleDelete", args, args, reply); done {
+		return err
+	}
+
+	defer metrics.MeasureSince([]string{"acl", "bindingrule", "delete"}, time.Now())
+
+	// Verify token is permitted to modify ACLs
+	if rule, err := a.srv.ResolveToken(args.Token); err != nil {
+		return err
+	} else if rule == nil || !rule.ACLWrite() {
+		return acl.ErrPermissionDenied
+	}
+
+	_, rule, err := a.srv.fsm.State().ACLBindingRuleGetByID(nil, args.BindingRuleID)
+	if err != nil {
+		return err
+	}
+
+	if rule == nil {
+		return nil
+	}
+
+	req := structs.ACLBindingRuleBatchDeleteRequest{
+		BindingRuleIDs: []string{args.BindingRuleID},
+	}
+
+	resp, err := a.srv.raftApply(structs.ACLBindingRuleDeleteRequestType, &req)
+	if err != nil {
+		return fmt.Errorf("Failed to apply binding rule delete request: %v", err)
+	}
+
+	if respErr, ok := resp.(error); ok {
+		return respErr
+	}
+
+	*reply = true
+
+	return nil
+}
+
+func (a *ACL) BindingRuleList(args *structs.ACLBindingRuleListRequest, reply *structs.ACLBindingRuleListResponse) error {
+	if err := a.aclPreCheck(); err != nil {
+		return err
+	}
+
+	if done, err := a.srv.forward("ACL.BindingRuleList", args, args, reply); done {
+		return err
+	}
+
+	if rule, err := a.srv.ResolveToken(args.Token); err != nil {
+		return err
+	} else if rule == nil || !rule.ACLRead() {
+		return acl.ErrPermissionDenied
+	}
+
+	return a.srv.blockingQuery(&args.QueryOptions, &reply.QueryMeta,
+		func(ws memdb.WatchSet, state *state.Store) error {
+			index, rules, err := state.ACLBindingRuleList(ws, args.AuthMethod)
+			if err != nil {
+				return err
+			}
+
+			reply.Index, reply.BindingRules = index, rules
+			return nil
+		})
+}
+
+func (a *ACL) AuthMethodRead(args *structs.ACLAuthMethodGetRequest, reply *structs.ACLAuthMethodResponse) error {
+	if err := a.aclPreCheck(); err != nil {
+		return err
+	}
+
+	if done, err := a.srv.forward("ACL.AuthMethodRead", args, args, reply); done {
+		return err
+	}
+
+	if rule, err := a.srv.ResolveToken(args.Token); err != nil {
+		return err
+	} else if rule == nil || !rule.ACLRead() {
+		return acl.ErrPermissionDenied
+	}
+
+	return a.srv.blockingQuery(&args.QueryOptions, &reply.QueryMeta,
+		func(ws memdb.WatchSet, state *state.Store) error {
+			index, method, err := state.ACLAuthMethodGetByName(ws, args.AuthMethodName)
+
+			if err != nil {
+				return err
+			}
+
+			reply.Index, reply.AuthMethod = index, method
+			return nil
+		})
+}
+
+func (a *ACL) AuthMethodSet(args *structs.ACLAuthMethodSetRequest, reply *structs.ACLAuthMethod) error {
```

Comment:
Throwing the comment here but I think it could apply to all the binding rule and Auth method crud endpoints:

We could potentially check if token replication is enabled and return errors indicating that its a dependency of all the secure intro stuff.

Reply:
Done. And I added two test scenarios to verify:
* auth method, binding rule, login, logout endpoints fail with the informational error when token replication is disabled
* auth method, binding rule, login, logout endpoints only operate on the local datacenter (technically there was a bug in most of these endpoints that forwarded the writes to the primary that i hadn't noticed)